### ISSUE TITLE
Detail design of architecture

### ISSUE THREAD

**Issue by mrjwood** [mrjwood/AgileManagementAssistant#1]

**Comment by mrjwood** [comment]
MVP architecture
My minimal viable product (MVP) is a standalone web hook so for this I do not need to consider the architecture too much. For a Webhook, it makes sense to use pure Node.js as it is very lightweight and for response time the “V8 VM JIT compiler of chrome that is extraordinary faster than Python” (Sehar, 2016) makes it the preferred option for this type of application. On the other hand, if we consider the machine learning (ML) side of my application then it could be argued that Python is the better option due to its history in the field and number of libraries supported for it. 

Node.js/JavaScript (JS) have got their own ML libraries, however, they have been restricted to the speed of JavaScript. Python libraries such as Tensorflow offer GPU support which can perform operations much quicker.  This year Google released DeepLearn.js, “an open source WebGL-accelerated JS library for machine learning”. By using WebGL the library can perform computations on the GPU which is one of the main arguments to favour other programming-languages for ML. Therefore, it seems that Node.js/JS are having ML options being made available which I could potentially use.

If the capabilities of the Node.js/JS libraries are not suitable for my application, I can instead use a python library and integrate it with my Node.js front. This can be done by making a child process to run the heavy python computation in the background, leaving the node.js app free to respond to the client.

In conclusion, Node.js is preferable for Webhooks and I have options for the ML side of the application when I get to that point. Therefore, I will start with simply building my node.js up first.
